### PR TITLE
feat(18842): Implement connectivity controls between the different types of node

### PR DESCRIPTION
**Pull request by vanch3d**
See https://hivemq.kanbanize.com/ctrl_board/57/cards/18842/details/

This PR complements the work done on the new visual language for policies by adding and visualising constraints between elements on the canvas.

The PR also fixes a few styling issues with the nodes.

### Design
- The two types of handles on nodes have been made more visible to highlight `target` (circle) and `source` (square) 
- Connecting two nodes can only be done between a source and a target. Drag-and-drop can be used from any direction but the created link will be directional.
- A source node can only be connected to a valid target. Each node defined its own validity (e.g. `Topic Filter` connected to `Data Policy` only)
- Some nodes (e.g. `Data Policy` or `Operation`) also have cardinality constraints and only allow one link to another destination. Each handle (e.g. `onSuccess` or `onError`) has its constraints. 
- When dragging a link from one node to another, the validity of the targetted handle will be shown; if invalid, the connection cannot be established 

### Before
Any node could be connected to any other node

screenshot-localhost_3000-2024 01 24-13_24_05

### After
Only valid connections can be established

screenshot-localhost_3000-2024 01 24-13_26_40

Visual feedback for the validity of an attempted connection is now provided 

screenshot-localhost_3000-2024 01 24-11_39_27

screenshot-localhost_3000-2024 01 24-11_39_42

screenshot-localhost_3000-2024 01 24-11_39_58

### CODE REVIEW

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/ClientFilterNode.tsx [hunk 1]
```diff
@@ -18,25 +18,23 @@ export const ClientFilterNode: FC<NodeProps<ClientFilterData>> = (props) => {
       <NodeWrapper route={`node/${type}/${id}`} {...props}>
         <HStack>
           <VStack>
-            <Text> {t('workspace.nodes.type', { context: type })}</Text>
+            <Text data-testid={'node-title'}> {t('workspace.nodes.type', { context: type })}</Text>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/ClientFilterNode.tsx:21]: suggestion(non-blocking): small little change on the attribute definitiona

```suggestion
            <Text data-testid="node-title"> {t('workspace.nodes.type', { context: type })}</Text>
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/ClientFilterNode.tsx [hunk 2]
```diff
@@ -18,25 +18,23 @@ export const ClientFilterNode: FC<NodeProps<ClientFilterData>> = (props) => {
...
           </VStack>
         </HStack>
-        <VStack ml={6}>
+        <VStack ml={6} data-testid={'node-model'}>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/ClientFilterNode.tsx:24]: suggestion(non-blocking): small little change on the attribute definitiona

```suggestion
        <VStack ml={6} data-testid="node-model">
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/CustomHandle.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  isConnectable?: boolean | number
+}
+
+export const CustomHandle: FC<CustomHandleProps> = (props) => {
+  const { nodes, edges } = useDataHubDraftStore()
+  const nodeId = useNodeId()
+
+  const isHandleConnectable = useMemo(() => {
+    if (typeof props.isConnectable === 'number') {
+      const node = nodes.find((e) => e.id === nodeId)
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/CustomHandle.tsx:16]: question(non-blocking): I'm a big fan of preventing abbreviations and use the sigular of a list
It becomes easier to parse. As an example I first thought it was an `event` then `error` then I saw it's a filter.

```suggestion
      const node = nodes.find((node) => node.id === nodeId)
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/CustomHandle.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      if (node) {
+        const connectedEdges = getConnectedEdges([node], edges)
+
+        const toHandle = connectedEdges.filter((e) => {
+          const otherEnd = props.type === 'source' ? e.sourceHandle : e.targetHandle
+          return otherEnd === props.id
+        })
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/CustomHandle.tsx:23]: question: similar to the previous comment

>question(non-blocking): I'm a big fan of preventing abbreviations and use the sigular of a list
It becomes easier to parse. As an example I first thought it was an event then error then I saw it's a filter.

```suggestion
        const toHandle = connectedEdges.filter((edge) => {
          const otherEnd = props.type === 'source' ? edge.sourceHandle : edge.targetHandle
          return otherEnd === props.id
        })
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/DataPolicyNode.spec.cy.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      })
+  })
+
+  it('should be accessible', () => {
+    cy.injectAxe()
+    cy.mountWithProviders(mockReactFlow(<DataPolicyNode {...MOCK_NODE_DATA_POLICY} />))
+    cy.checkAccessibility()
+    cy.percySnapshot('Component: DataHub - BehaviorPolicyNode')
+  })
+})
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/DataPolicyNode.spec.cy.tsx:71]: praise: very nice test!

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/DataPolicyNode.tsx [hunk 6]
```diff
@@ -17,38 +17,38 @@ export const DataPolicyNode: FC<NodeProps<DataPolicyData>> = (props) => {
       <NodeWrapper route={`node/${DataHubNodeType.DATA_POLICY}/${id}`} {...props}>
         <HStack>
           <NodeIcon type={DataHubNodeType.DATA_POLICY} />
-          <Text> {t('workspace.nodes.type', { context: type })}</Text>
+          <Text data-testid={'node-title'}> {t('workspace.nodes.type', { context: type })}</Text>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/DataPolicyNode.tsx:20]: suggestion: let's assign the attribute without the evaluation

```suggestion
          <Text data-testid="node-title"> {t('workspace.nodes.type', { context: type })}</Text>
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/DataPolicyNode.tsx [hunk 7]
```diff
@@ -17,38 +17,38 @@ export const DataPolicyNode: FC<NodeProps<DataPolicyData>> = (props) => {
...
         </HStack>
-        <VStack ml={6} alignItems={'flex-end'}>
+        <VStack ml={6} alignItems={'flex-end'} data-testid={'node-model'}>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/DataPolicyNode.tsx:22]: suggestion: let's assign the attribute without the evaluation

```suggestion
        <VStack ml={6} alignItems={'flex-end'} data-testid="node-model">
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/DataPolicyNode.tsx [hunk 8]
```diff
@@ -17,38 +17,38 @@ export const DataPolicyNode: FC<NodeProps<DataPolicyData>> = (props) => {
...
+      <CustomHandle
         type="source"
         position={Position.Right}
         id={DataPolicyData.Handle.ON_SUCCESS}
+        className={DataPolicyData.Handle.ON_SUCCESS}
         style={{
           top: `calc(var(--chakra-space-3) + 10px)`,
-          background: 'green',
-          ...styleSourceHandle,
+          // background: 'green',
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/DataPolicyNode.tsx:38]: suggestion: found a leftover, not sure if it needs to be added

```suggestion
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/OperationNode.spec.cy.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        expect((attr as unknown as string).endsWith('target')).to.be.true
+      })
+  })
+
+  // it('should be accessible', () => {
+  //   cy.injectAxe()
+  //   cy.mountWithProviders(mockReactFlow(<OperationNode {...MOCK_NODE_OPERATION} />))
+  //   cy.checkAccessibility()
+  //   cy.percySnapshot('Component: DataHub - OperationNode')
+  // })
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/OperationNode.spec.cy.tsx:94]: suggestion: add the commented out test

Or Is that one flaky?

```suggestion
it('should be accessible', () => {
  cy.injectAxe()
  cy.mountWithProviders(mockReactFlow(<OperationNode {...MOCK_NODE_OPERATION} />))
  cy.checkAccessibility()
  cy.percySnapshot('Component: DataHub - OperationNode')
})
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/OperationNode.tsx [hunk 10]
```diff
@@ -1,34 +1,44 @@
...
 
+  const model = useMemo(() => {
+    if (!action) return undefined
+
+    return typeof action === 'string'
+      ? ({
+          functionId: action,
+        } as FunctionDefinition)
+      : action
+  }, [action])
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/OperationNode.tsx:24]: suggestion: fully valid, but a bit hard to parse. WDYT about this one?

```suggestion
  const model = useMemo(() => {
    if (!action) return undefined
    
    if (typeof action === 'string') {
      return {
        functionId: action,
      } satisfies FunctionDefinition
    }

    return action
  }, [action])
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/OperationNode.tsx [hunk 11]
```diff
@@ -1,34 +1,44 @@
...
+
   return (
     <>
       <NodeWrapper route={`node/${DataHubNodeType.OPERATION}/${id}`} {...props}>
         <HStack>
           <NodeIcon type={DataHubNodeType.OPERATION} />
-          <Text> {t('workspace.nodes.type', { context: type })}</Text>
+          <Text data-testid={'node-title'}> {t('workspace.nodes.type', { context: type })}</Text>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/OperationNode.tsx:31]: suggestion: sorry for annoying you with this one :D

```suggestion
          <Text data-testid="node-title"> {t('workspace.nodes.type', { context: type })}</Text>
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/OperationNode.tsx [hunk 12]
```diff
@@ -1,34 +1,44 @@
...
           <VStack>
-            <Text>{action?.functionId || '< none >'}</Text>
+            <Text data-testid={'node-model'}>{model?.functionId || t('error.noSet.select')}</Text>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/OperationNode.tsx:33]: suggestion: sorry for annoying you with this one :D

```suggestion
            <Text data-testid="node-model">{model?.functionId || t('error.noSet.select')}</Text>
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/SchemaNode.tsx [hunk 13]
```diff
@@ -1,28 +1,29 @@
...
+  const { t } = useTranslation('datahub')
+  const { id, data, type } = props
 
   return (
     <>
       <NodeWrapper route={`node/${DataHubNodeType.SCHEMA}/${id}`} {...props}>
         <HStack>
           <NodeIcon type={DataHubNodeType.SCHEMA} />
-          <Text>Schema</Text>
+          <Text data-testid={'node-title'}> {t('workspace.nodes.type', { context: type })}</Text>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/SchemaNode.tsx:20]: suggestion: maybe we can align eslint to do that automatically

```suggestion
          <Text data-testid="node-title"> {t('workspace.nodes.type', { context: type })}</Text>
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/TransitionNode.spec.cy.tsx [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+  beforeEach(() => {
+    cy.viewport(400, 400)
+  })
+
+  it('should render properly', () => {
+    cy.mountWithProviders(mockReactFlow(<TransitionNode {...MOCK_NODE_TRANSITION} selected={true} />))
+    cy.getByTestId(`node-title`).should('contain.text', 'Transition')
+    cy.getByTestId(`node-model`).should('contain.text', '< not set >')
+
+    // TODO[NVL] Create a PageObject and generalise the selectors
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/TransitionNode.spec.cy.tsx:28]: question: What is the process on TODOs? Is that being added to the backlog

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/TransitionNode.tsx [hunk 15]
```diff
@@ -17,14 +17,14 @@ export const TransitionNode: FC<NodeProps<TransitionData>> = (props) => {
       <NodeWrapper route={`node/${DataHubNodeType.TRANSITION}/${id}`} {...props}>
         <HStack>
           <NodeIcon type={DataHubNodeType.TRANSITION} />
-          <Text> {t('workspace.nodes.type', { context: type })}</Text>
+          <Text data-testid={'node-title'}> {t('workspace.nodes.type', { context: type })}</Text>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/TransitionNode.tsx:20]:
```suggestion
          <Text data-testid="node-title"> {t('workspace.nodes.type', { context: type })}</Text>
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/TransitionNode.tsx [hunk 16]
```diff
@@ -17,14 +17,14 @@ export const TransitionNode: FC<NodeProps<TransitionData>> = (props) => {
...
           <VStack>
-            <Text>{data.type || '< none >'}</Text>
+            <Text data-testid={'node-model'}>{data.type || t('error.noSet.select')}</Text>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/TransitionNode.tsx:22]:
```suggestion
            <Text data-testid="node-model">{data.type || t('error.noSet.select')}</Text>
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/ValidatorNode.tsx [hunk 17]
```diff
@@ -16,16 +16,17 @@ export const ValidatorNode: FC<NodeProps<ValidatorData>> = (props) => {
       <NodeWrapper route={`node/${DataHubNodeType.VALIDATOR}/${id}`} {...props}>
         <HStack>
           <NodeIcon type={DataHubNodeType.VALIDATOR} />
-          <Text w={'50%'}> {t('workspace.nodes.type', { context: type })}</Text>
-          <VStack>
-            <Text>{data.strategy}</Text>
+          <Text data-testid={'node-title'} w={'50%'}>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/ValidatorNode.tsx:19]:
```suggestion
          <Text data-testid="node-title" w="50%">
```

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/ValidatorNode.tsx [hunk 18]
```diff
@@ -16,16 +16,17 @@ export const ValidatorNode: FC<NodeProps<ValidatorData>> = (props) => {
...
+            {t('workspace.nodes.type', { context: type })}
+          </Text>
+          <VStack data-testid={'node-model'}>
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/ValidatorNode.tsx:22]:
```suggestion
          <VStack data-testid="node-model">
```

**Review by h2xd — APPROVED** [review]
LGTM! 🐊

**Comment by h2xd** [comment]
Added couple of comments and suggestions mostly on code style. Looks great overall on the first view. Very well done!

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/CustomHandle.tsx (hunk 3 above)
Comment (vanch3d) [reply hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/CustomHandle.tsx:16]: Lol, that's `e` for element!
It's interesting how a good IDE changes practices, as the implicit typing is always visualised on my editors so I don't have the confusion for the small internals like these.

Screenshot 2024-01-25 at 09 54 58

Point taken will pay more attention

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/ClientFilterNode.tsx (hunk 1 above)
Comment (vanch3d) [reply hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/ClientFilterNode.tsx:21]: Lots of them around. I just realised I didn't configure eslint/prettier for `react/jsx-curly-brace-presence`.
Will fix the config and run a cleaning operation in a separate ticket

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/TransitionNode.spec.cy.tsx (hunk 14 above)
Comment (vanch3d) [reply hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/TransitionNode.spec.cy.tsx:28]: > question: What is the process on TODOs? Is that being added to the backlog

If they come with a ticket number (`TODO[18458]`), they are in the backlog and will be removed when solved.
The others should be added to the backlog and processed... at some point :-)
I usually do that when I come across one that I am in a position to address.

File: hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/CustomHandle.tsx (hunk 3 above)
Comment (h2xd) [reply hivemq-edge/src/frontend/src/extensions/datahub/components/nodes/CustomHandle.tsx]: Ah that make sense! I like to review within GitHub. That one does sadly not include the Type inference. Thank you for taking care 👍
